Products/DataCollector/ApplyDataMap/datamaputils.py

```diff
@@ -158,26 +158,63 @@ def _decode_value(value, obj):
     # This looks confusing, and it is. The scenario is:
     #   A collector gathers some data as a raw byte stream,
     #   but really it has a specific encoding specified by
-    #   by the zCollectorDecoding zProperty. Say, latin-1 or
+    #   the zCollectorDecoding zProperty. Say, latin-1 or
     #   utf-16, etc. We need to decode that byte stream to get
     #   back a UnicodeString object. But, this version of Zope
     #   doesn't like UnicodeString objects for a variety of
     #   fields, such as object ids, so we then need to convert
     #   that UnicodeString back into a regular string of bytes,
     #   and for that we use the system default encoding, which
     #   is now utf-8.
+
     try:
         codec = obj.zCollectorDecoding
     except AttributeError:
         codec = sys.getdefaultencoding()
 
-    value = value.decode(codec)
-    value = value.encode(sys.getdefaultencoding())
+    if not codec:
+        log.warn(
+            "zCollectorDecoding is not defined! It can cause problems "
+            "with the modeling and even can crash it, therefore specify "
+            "the correct data encoding as a value of "
+            "zCollectorDecoding.")
+
+        obj.zCollectorDecoding = 'utf-8'
+        codec = obj.zCollectorDecoding
+
+    try:
+        value = value.decode(codec)
+    except UnicodeDecodeError as ex1:
+        log.warn("The string %s cannot be correctly decoded using '%s' and "
+                 "causes error: %s! Please verify and specify the correct "
+                 "data encoding as a value of zCollectorDecoding.", value,
+                 codec, ex1)
+        try:
+            obj.zCollectorDecoding = 'latin-1'
+            codec = obj.zCollectorDecoding
+            value = value.decode(codec)
+        except UnicodeDecodeError as ex2:
+            log.warn("The string %s cannot be correctly decoded using '%s' "
+                     "and causes error: %s! Please verify and specify the "
+                     "correct data encoding as a value of zCollectorDecoding",
+                     value, codec, ex2)
+
+            value = value.decode(codec, errors='ignore')
+
+            log.warn("The string %s was decoded to %s ignoring symbols "
+                     "which is impossible to decode into '%s' ",
+                     value, codec, codec)
+
+    encoder = sys.getdefaultencoding()
+    if not encoder:
+        encoder = 'utf-8'
+    value = value.encode(encoder)
+
```

Comment:
Now that I've had some time to think about this issue and review your changes in response to our original conversation, I think a different strategy will have some benefits.

After relearning that the `decode` method has an `errors` argument, I think the strategy of trying to guess the encoding is more work than necessary to accomplish our goal.  The goal of the `_decode_value` function is to normalize strings to the system's default encoding.  For troubleshooting purposes, a side goal would be to log when specifying `zCollectorDecoding` would be helpful for decoding the value.  Another thing to note is that we've configured `sys.getdefaultencoding` to return "utf-8" as the default codec (have a look at /opt/zenoss/lib/python2.7/sitecustomize.py in an image).

So, maybe an implementation like the following will suffice:
```
    try:
        value = value.decode(codec)
    except UnicodeDecodeError as ex:
        value = value.decode(codec, errors="ignore")
        log.warn(
            "Unable to decode string using codec '%s'.  "
            "Using a modified string without errors.  "
            "Please set zCollectorDecoding to the correct codec. "
            "object=%r modified-value='%s' error=%s",
            codec, obj, value, ex,
        )
            
    return value.encode(sys.getdefaultencoding())
```